[RnD-EasyPurchase-iOS][478621]Implement Payment Controller

shamin26 opened the pull request:
- Defined the protocol for Transaction Controller.
- Defined PaymentController to handle & process transactions.
- Added detailed comments to improve code clarity.

Ref : https://redmine.bjitgroup.com/redmine/issues/478621

shamin26 committed:
[RnD-EasyPurchase-iOS][478621]Implement Payment Controller

- Defined the protocol for Transaction Controller.
- Defined PaymentController to handle & process transactions.
- Added detailed comments to improve code clarity.

Ref : https://redmine.bjitgroup.com/redmine/issues/478621

RejaulHasan commented on Sources/EasyPurchase/PaymentsController.swift:
```diff
@@ -0,0 +1,96 @@
+}
+
+// Implementation of the transaction controller
+class PaymentsController: TransactionController {
```
If no one inharit this class then make it final. Like
final class PaymentsController{}

shamin26 replied:
Noted.

RejaulHasan commented on Sources/EasyPurchase/PaymentsController.swift:
```diff
@@ -0,0 +1,96 @@
+
+    var payments: [Payment] = []          // Array to hold pending payments
+    var unhandledTransactions: [SKPaymentTransaction] = [] // Array to hold unhandled transactions
+
```
If you do not use those property outside of the class then make it private property.

shamin26 replied:
Noted.

RejaulHasan commented on Sources/EasyPurchase/PaymentsController.swift:
```diff
@@ -0,0 +1,96 @@
+import Foundation
+import StoreKit
+
+// Structure representing a payment
```
Please write comment like below link. It will help to read.
-> https://rejaul-hasan.medium.com/document-your-swift-code-6eb1223c53d6

shamin26 replied:
Noted.

RejaulHasan commented on Sources/EasyPurchase/PaymentsController.swift:
```diff
@@ -0,0 +1,96 @@
+class PaymentsController: TransactionController {
+
+    var payments: [Payment] = []          // Array to hold pending payments
+    var unhandledTransactions: [SKPaymentTransaction] = [] // Array to hold unhandled transactions
```
var unhandledTransactions: [SKPaymentTransaction] = []

-> Why store this property as we do not need it and it's a stateless.

shamin26 replied:
Noted

RejaulHasan commented on Sources/EasyPurchase/PaymentsController.swift:
```diff
@@ -0,0 +1,96 @@
+            return false
+
+        case .restored:
+            // Transaction was restored (e.g., for a previously purchased non-consumable)
```
Not handling it. Don't we need it?

shamin26 replied:
Restore feature is not implemented yet. Should I handle it at this stage?

FaiazRahman786 replied:
As PaymentController is only responsible for the deal with the purchased product, _`func processTransaction()->Bool`_ should consider the` .purchased `case only.
1. call _`findPaymentIndex()`_ to verify if the global variable `payment[]` (which was assigned by `paymentQueueController`) carries the passed transaction from this function. If not found return.
2. If found, check if the state is `.purchased`, if true -> create the essential object and pass it to the callback.
3. finish the transaction if no need to download anything and remove the payment.
4. return true.

RejaulHasan commented on Sources/EasyPurchase/PaymentsController.swift:
```diff
@@ -0,0 +1,96 @@
+
+        case .deferred:
+            // Transaction is in a deferred state (e.g., for family sharing)
+            // Handle as needed based on your app's requirements
```
Not handling it. Don't we need it?

shamin26 replied:
We are just considering purchasing, purchased and failed state for this commit. Deferred feature is not implemented yet. Should I handle it at this stage?

RejaulHasan reviewed (changes requested):
I attached the link also. Have a look. Thanks.

FaiazRahman786 commented on Sources/EasyPurchase/PaymentsController.swift:
```diff
@@ -0,0 +1,96 @@
+
+        // Store unhandled transactions in the property
+        self.unhandledTransactions = unhandledTransactions
+
```
We may use `unhandledTransactions` locally in this function but do not need it as global var, as we never used it elsewhere.

shamin26 committed:
[IAP][Func][Review-Feedback]: Implement Payment Controller

- Fixed the class scope issue
- Fixed the stored property issue
- Fixed the comment convention issue

Ref : https://redmine.bjitgroup.com/redmine/issues/478621

sadat-399 reviewed (approved):
LGTM

TulonBJIT reviewed (approved):
It looks ok on my side.

TulonBJIT committed:
Merge branch 'develop' of ssh://github.com/TulonBJIT/EasyPurchase into Task_478621
